tools/opam-updates.py

```diff
@@ -0,0 +1,326 @@
+
+    def __init__(self, name, cur=None, new=None):
+        self.name = name
+        if cur is None and new is None:
+            raise ValueError("At least one of cur and new has to be present.")
+        self.cur = cur
+        self.new = new
+
+    @property
+    def github_api_url(self):
+        "GitHub API url for the opam package folder content."
+        path = "packages/{name}/{name}.{version}".format(
+            name=self.name,
+            version=self.new if self.new is not None else self.cur)
+        return "https://api.github.com/repos/ocaml/opam-repository/contents/{}".format(path)
+
+    @property
+    def current_name(self):
+        "Return name.current_version"
+        return "{}.{}".format(self.name, self.cur)
+
+    @property
+    def new_name(self):
+        "Return name.new_version"
+        return "{}.{}".format(self.name, self.new)
+
+    def current_path(self, prefix):
+        "Path of the current version based on [prefix]"
+        return os.path.join(prefix, self.current_name)
+
+    def new_path(self, prefix):
+        "Path of the new version based on [prefix]"
+        return os.path.join(prefix, self.new_name)
+
+    def path(self, prefix):
+        """
+        Path of the new version based on [prefix] if present,
+        otherwise current path.
+        """
+        if self.new is not None:
+            return self.new_path(prefix)
+        return self.current_path(prefix)
+
+
+def download(session, url, file_path):
+    "Download the file at [url] in [file_path]."
+    r = session.get(url)
+    r.raise_for_status()
+    with open(file_path, 'wb') as f:
+        f.write(r.content)
+        f.flush()
+
+
+def update_files(session, auth, package, prefix):
+    """
+    Lookup the opam files for the [package], and update them in the
+    appropriate subfolder in [prefix].
+    """
+    base_url = package.github_api_url
+
+    r = session.get(base_url, auth=auth)
+    r.raise_for_status()
+    entries = r.json()
+
+    for entry in entries:
+        if entry["type"] == "dir":
+            continue
+        new_path = os.path.join(package.path(prefix), entry["name"])
+        download(session, entry["download_url"], new_path)
+
+    if any(entry["type"] == "dir"
+           and entry["name"] == "files"
+           for entry in entries):
+        r = session.get(os.path.join(base_url, "files"), auth=auth)
+        r.raise_for_status()
+        entries = r.json()
+        prefix = os.path.join(package.path(prefix), "files")
+        os.makedirs(prefix, exist_ok=True)
+
+        for entry in entries:
+            if entry["type"] == "dir":
+                continue
+            new_path = os.path.join(package.path(prefix), entry["name"])
+            download(session, entry["download_url"], new_path)
+
+
+def extract_upgrade(row):
+    "Extract upgraded packages from a row in the opam output."
+    name, cur, _, new = row.strip().split()[2:6]
+    return Package(name, cur, new)
+
+
+def extract_update(row):
```

Comment:
What is the difference between an upgraded and an updated package? Is this the "reinstall"?

Reply:
upgrade: fresh new version
update: something changed upstream but the version is unchanged